### PR TITLE
units.Quantity support for scipy.special ufuncs

### PR DESCRIPTION
**Pull request by kyleaoman**
Still lots to do, but the first few scipy.special ufuncs now work with quantities. 

EDIT: Address part of #6390

### CODE REVIEW

File: astropy/units/quantity_helper.py [hunk 1]
```diff
@@ -636,3 +643,273 @@ def check_output(output, unit, inputs, function=None):
         raise TypeError("Arguments cannot be cast safely to inplace "
                         "output with dtype={0}".format(output.dtype))
     return output
+
+# helpers for ufuncs in scipy.special
+# available ufuncs in this module are at 
+# https://docs.scipy.org/doc/scipy/reference/special.html
+
+try:
+    import scipy.special
```
Comment (mhvk) [inline astropy/units/quantity_helper.py]: Maybe `from scipy import special as <something-short>` so the text below is shorter as well?

File: astropy/units/quantity_helper.py [hunk 2]
```diff
@@ -83,18 +82,26 @@ def get_converter(from_unit, to_unit):
 
 # ufuncs handled as special cases
 
-UFUNC_HELPERS[np.sqrt] = lambda f, unit: (
+helper_sqrt = lambda f, unit: (
```
[the remark this answers is not in the document]
Comment (kyleaoman) [reply astropy/units/quantity_helper.py]: Will do, I named a few lambdas I wanted to re-use, but for consistency they should probably all just be def's. I will probably move all of them to the top now that some need using in multiple places.

File: astropy/units/quantity_helper.py (hunk 1 above)
Comment (kyleaoman) [reply astropy/units/quantity_helper.py]: Sure, wasn't sure what the preferred style here was.

**Comment by kyleaoman** [comment]
A large fraction of unit tests in this module feed a function some output then check the result against some output. Before I go ahead and ~double the number of these, I wonder whether something like what I illustrate below would make sense. Exactly the same functionality but, imo, easier to read and maintain. The results of failed tests are also still easy to parse.

```python
from collections import namedtuple
testcase = namedtuple('testcase', ['f', 'q_in', 'q_out'])

class TestQuantityTrigonometricFuncs:
    """
    Test trigonometric functions
    """
    @pytest.mark.parametrize('tc', (
        testcase(
            f=np.sin, 
            q_in=(30. * u.degree, ), 
            q_out=(0.5*u.dimensionless_unscaled, )
        ),
        testcase(
            f=np.sin, 
            q_in=(np.array([0., np.pi / 4., np.pi / 2.]) * u.radian, ),
            q_out=(np.array([0., 1. / np.sqrt(2.), 1.]) * u.dimensionless_unscaled, )
        ),
        testcase(
            f=np.arcsin,
            q_in=(np.sin(30. * u.degree), ),
            q_out=(np.radians(30.) * u.radian, )
        )
    ))
    def test_testcase(self, tc):
        results = tc.f(*tc.q_in)
        results = (results, ) if type(results) != tuple else results
        for result, expected in zip(results, tc.q_out):
            assert result.unit == expected.unit
            assert_allclose(result.value, expected.value)
```

**Comment by mhvk** [comment]
@kyleaoman - I very much like your reorganization of `quantity_helpers` - much clearer to have the definitions explicit.

I also rather like your suggestion for re-organizing the test cases, but realize it is a lot of work...

**Comment by mhvk** [comment]
@kyleaoman - I had lost track of where this was, but felt it would be very nice to have this in 3.0, so went ahead and did a rebase (rather messy...) as well as some clean-up. The one thing that deserves some comment is that I removed the addition to `UNSUPPORT_UFUNCS` of all the functions in `scipy.special` that we do not yet cover - the reason is that this set is meant for functions that we really cannot support - because they make no sense with quantities - we should leave out things we haven't decided on yet. The only real difference this makes is the error message - and we want to encourage users to report to us if they would like a function supported.

**Comment by kyleaoman** [comment]
Looks pretty good. I might do some more next year in terms of adding more supported functions and cleaning up the tests, I have an occasional long train commute that I'm thinking will give me a good excuse.

**Comment by mhvk** [comment]
Merging since the failed test is just due to a different issue (fixed in #7027).

**Comment by mhvk** [comment]
Thanks, @kyleaoman!

**Comment by namurphy** [comment]
This is wonderful - thank you!

**Comment by bsipocz** [comment]
@mhvk - Could you please either add some what's new text to http://docs.astropy.org/en/latest/whatsnew/3.0.html#what-s-new-in-astropy-3-0, or drop the label.

**Comment by mhvk** [comment]
Ah, yes.  See #7138

**Pull request merged**